File: benchmark/ts/run-all.ts
/**
 * Master benchmark runner for NumJS (TypeScript/WASM).
 * Runs all benchmark categories and outputs results to JSON.
 */

import { writeFileSync, mkdirSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

import { runReduceBenchmarks } from './reduce/aggregations.js';
import { runCoreBenchmarks } from './core/creation.js';
import { runUfuncBenchmarks } from './ufunc/arithmetic.js';
import { runLinalgBenchmarks } from './linalg/operations.js';
import { runRandomBenchmarks } from './random_bench/distributions.js';
import { CATEGORY_NAMES } from './lib/config.js';
import type { CategoryResults } from './lib/types.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Parse command line arguments
const args = process.argv.slice(2);
const categoryArg = args.find(a => a.startsWith('--category='));
const selectedCategory = categoryArg?.split('=')[1];

async function main() {
  console.log('='.repeat(60));
  console.log('NumJS Benchmark Suite');
  console.log('='.repeat(60));

  const categories: CategoryResults[] = [];

  // Run selected or all categories
  const runCategory = async (
    id: string,
    runner: () => Promise<any[]>
  ) => {
    if (selectedCategory && selectedCategory !== id) {
      return;
    }
    console.log(`\n[${CATEGORY_NAMES[id] || id}]`);
    const operations = await runner();
    categories.push({
      id,
      name: CATEGORY_NAMES[id] || id,
      operations,
    });
  };

  await runCategory('reduce', runReduceBenchmarks);
  await runCategory('core', runCoreBenchmarks);
  await runCategory('ufunc', runUfuncBenchmarks);
  await runCategory('linalg', runLinalgBenchmarks);
  await runCategory('random', runRandomBenchmarks);

  // Output results
  const projectRoot = join(__dirname, '..', '..', '..');
  const outputDir = join(projectRoot, 'benchmark', 'results', 'numjs');
  mkdirSync(outputDir, { recursive: true });

  // Write individual category files
  for (const category of categories) {
    const outputFile = join(outputDir, `${category.id}.json`);
    writeFileSync(outputFile, JSON.stringify(category, null, 2));
    console.log(`\nResults saved: ${outputFile}`);
  }

  console.log('\n' + '='.repeat(60));
  console.log('NumJS benchmarks complete!');
  console.log('='.repeat(60));
}

main().catch((err) => {
  console.error('Benchmark failed:', err);
  process.exit(1);
});

File: packages/numwasm/benchmark/python/run_all.py
#!/usr/bin/env python3
"""
Master benchmark runner for NumPy.
Runs all benchmark categories and outputs results to JSON.
"""

import sys
import os
import json
from pathlib import Path

# Add the benchmark directory to the path
sys.path.insert(0, str(Path(__file__).parent.parent))

from python.reduce.aggregations import run_reduce_benchmarks
from python.core.creation import run_core_benchmarks
from python.ufunc.arithmetic import run_ufunc_benchmarks
from python.linalg.operations import run_linalg_benchmarks
from python.random_bench.distributions import run_random_benchmarks
from python.lib.config import CATEGORY_NAMES


def main():
    # Parse command line arguments
    selected_category = None
    for arg in sys.argv[1:]:
        if arg.startswith('--category='):
            selected_category = arg.split('=')[1]

    print('=' * 60)
    print('NumPy Benchmark Suite')
    print('=' * 60)

    categories = []

    def run_category(cat_id: str, runner):
        if selected_category and selected_category != cat_id:
            return
        print(f"\n[{CATEGORY_NAMES.get(cat_id, cat_id)}]")
        operations = runner()
        categories.append({
            'id': cat_id,
            'name': CATEGORY_NAMES.get(cat_id, cat_id),
            'operations': operations,
        })

    run_category('reduce', run_reduce_benchmarks)
    run_category('core', run_core_benchmarks)
    run_category('ufunc', run_ufunc_benchmarks)
    run_category('linalg', run_linalg_benchmarks)
    run_category('random', run_random_benchmarks)

    # Output results
    project_root = Path(__file__).parent.parent.parent
    output_dir = project_root / 'benchmark' / 'results' / 'numpy'
    output_dir.mkdir(parents=True, exist_ok=True)

    # Write individual category files
    for category in categories:
        output_file = output_dir / f"{category['id']}.json"
        with open(output_file, 'w') as f:
            json.dump(category, f, indent=2)
        print(f"\nResults saved: {output_file}")

    print('\n' + '=' * 60)
    print('NumPy benchmarks complete!')
    print('=' * 60)


if __name__ == '__main__':
    main()
